### DISCUSSION TITLE
Zustand with typescript, slicing and persisting middleware

### DISCUSSION THREAD

**Discussion by ghost** [pmndrs/zustand#2027] [Q&A] [answered]
Hey everyone,

I am trying to build a next.js app with Zustand as a state store. The application is built in typescript and I plan to have two separate stores with different kinds of data. I took the tutorial on slicing which you can find [here](https://github.com/pmndrs/zustand/blob/main/docs/guides/typescript.md#slices-pattern).

Below the guide is explained how to add middleware such a persist. According to the guide, I have to add persist right after the `StateCreator` in the `<>` braces.

I end up having the following code:

```
import { create, StateCreator } from 'zustand'

interface BearSlice {
    bears: number
    addBear: () => void
    eatFish: () => void
}
const createBearSlice: StateCreator<
    BearSlice & FishSlice,
    [["zustand/persist", BearSlice]],
    [],
    BearSlice
> = (set) => ({
    bears: 0,
    addBear: () => set((state) => ({ bears: state.bears + 1 })),
    eatFish: () => set((state) => ({ fishes: state.fishes - 1 })),
})

interface FishSlice {
    fishes: number
    addFish: () => void
}
const createFishSlice: StateCreator<
    BearSlice & FishSlice,
    [["zustand/persist", FishSlice]],
    [],
    FishSlice
> = (set) => ({
    fishes: 0,
    addFish: () => set((state) => ({ fishes: state.fishes + 1 })),
})

const useBoundStore = create<BearSlice & FishSlice>()((...a) => ({
    ...createBearSlice(...a),
    ...createFishSlice(...a),
}));
```

When compiling the TypeScript code, I get the following error (the affected lines are `...createBearSlice(...a)` and `...createBearSlice(...a)`:

```
TS2345: Argument of type 'StoreApi<BearSlice & FishSlice>' is not assignable to parameter of type 'Write<StoreApi<BearSlice & FishSlice>, StorePersist<BearSlice & FishSlice, FishSlice>>'.   Property 'persist' is missing in type 'StoreApi<BearSlice & FishSlice>' but required in type 'StorePersist<BearSlice & FishSlice, FishSlice>'.  persist.d.ts(98, 5): 'persist' is declared here.
```

Does somebody by accident know how to fix/implement state persisting when slicing the store while using TypeScript?

Thanks a lot!

Regards,

**Comment by dai-shi** [comment]
Here's the typescript playground: https://tsplay.dev/mpQdMW

**Reply by dai-shi** [reply, marked as answer]
It looks like this passes: https://tsplay.dev/m3VLLN

**Reply by ghost** [reply]
Yeah, thats true! Thanks a lot for helping here.
